drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,543 @@
+		(*blk++) = (*data++);
+	}
+
+	for (int i = 0; i < (dev_config->flash_cbs/2); i++) {
+		if (i < sizeof(address)) {
+			(*blk++) = (uint8_t)(((address >> (8 * i)) & 0xff));
+		} else {
+			(*blk++) = 0xff;
+		}
+	}
+}
+
+/*
+ * eeprom_emu_is_word_used: check if word is not empty
+ */
+static int eeprom_emu_is_word_used(const struct device *dev, const uint8_t *blk)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	for (int i = 0; i < dev_config->flash_cbs; i++) {
+		if ((*blk++) != 0xff) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+/*
+ * eeprom_emu_word_read: read basic word (cbs byte of data) item from address.
+ */
+static int eeprom_emu_word_read(const struct device *dev, off_t address,
+				uint8_t *data)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	const struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	uint8_t buf[dev_config->flash_cbs];
+	off_t direct_address;
+
+	direct_address = dev_data->page_offset + address;
+
+	/* Direct flash read */
+	eeprom_emu_flash_read(dev, direct_address, data, sizeof(buf));
+
+	/* Process changes written to flash */
+	off_t offset, ch_address;
+	bool mc1 = false, mc2 = false;
+
+	offset = dev_data->write_offset;
+	while (((!mc1) || (!mc2)) && (offset > dev_config->size)) {
+		offset -= sizeof(buf);
+		/* read the change */
+		eeprom_emu_flash_read(dev, dev_data->page_offset + offset, buf,
+				      sizeof(buf));
+		/* get the address from a change block */
+		ch_address = eeprom_emu_get_address(dev, buf);
+		if ((!mc1) && (ch_address == address)) {
+			memcpy(data, buf, sizeof(buf)/2);
+			mc1 = true;
+		}
+		if ((!mc2) && (ch_address == (address + sizeof(buf)/2))) {
+			memcpy(data + sizeof(buf)/2, buf, sizeof(buf)/2);
+			mc2 = true;
+		}
+	}
+	return 0;
+}
+
+/*
+ * eeprom_emu_compactor: start a new eeprom page and copy existing data to the
+ * new page. Invalidate the old page.
+ */
+static void eeprom_emu_compactor(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	off_t next_page_offset, rd_offset = 0;
+	uint8_t buf[dev_config->flash_cbs];
+
+	LOG_DBG("Compactor called for page at [0x%tx]",
+		(ptrdiff_t)dev_data->page_offset);
+
+	next_page_offset = dev_data->page_offset + dev_config->page_size;
+	if (next_page_offset >= dev_config->flash_size) {
+		next_page_offset = 0;
+	}
+
+	/* erase the new page */
+	eeprom_emu_flash_erase(dev, next_page_offset, dev_config->page_size);
+
+	/* copy existing data */
+	while (rd_offset < dev_config->size) {
+		eeprom_emu_word_read(dev, rd_offset, buf);
+		if (eeprom_emu_is_word_used(dev, buf)) {
+			eeprom_emu_flash_write(dev, next_page_offset +
+					       rd_offset, buf, sizeof(buf));
+		}
+		rd_offset += sizeof(buf);
+	}
+	/* invalidate the old page */
+	eeprom_emu_page_invalidate(dev, dev_data->page_offset);
+	dev_data->write_offset = dev_config->size;
+	dev_data->page_offset = next_page_offset;
```

Comment:
The `eeprom_emu_flash_erase`, the `eeprom_emu_flash_write` and `eeprom_emu_page_invalidate` may all fail here due to `flash_write_protection_set` yet the entire code will proceed , which will basically mean that compaction may not happen at all, yet the `dev_data` will be updated.